src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.cs

```diff
@@ -32,69 +33,14 @@ public partial class EventSourceGenerator : ISourceGenerator
-            EventSourceClass[]? eventSources = p.GetEventSourceClasses(receiver.CandidateClasses);
-
-            if (eventSources?.Length > 0)
-            {
-                Emitter? e = new Emitter(context);
-                e.Emit(eventSources, context.CancellationToken);
-            }
-        }
-
-        private sealed class SyntaxReceiver : ISyntaxReceiver
+        public void Initialize(IncrementalGeneratorInitializationContext context)
         {
-            private List<ClassDeclarationSyntax>? _candidateClasses;
-
-            public List<ClassDeclarationSyntax>? CandidateClasses => _candidateClasses;
-
-            public void OnVisitSyntaxNode(SyntaxNode syntaxNode)
-            {
-                // Only add classes annotated [EventSourceAutoGenerate] to reduce busy work.
-                const string EventSourceAttribute = "EventSourceAutoGenerateAttribute";
-                const string EventSourceAttributeShort = "EventSourceAutoGenerate";
-
-                // Only classes
-                if (syntaxNode is ClassDeclarationSyntax classDeclaration)
-                {
-                    // Check if has EventSource attribute before adding to candidates
-                    // as we don't want to add every class in the project
-                    foreach (AttributeListSyntax? cal in classDeclaration.AttributeLists)
-                    {
-                        foreach (AttributeSyntax? ca in cal.Attributes)
-                        {
-                            // Check if Span length matches before allocating the string to check more
-                            int length = ca.Name.Span.Length;
-                            if (length != EventSourceAttribute.Length && length != EventSourceAttributeShort.Length)
-                            {
-                                continue;
-                            }
+            IncrementalValuesProvider<EventSourceClass> eventSourceClasses =
+                context.SyntaxProvider
+                .CreateSyntaxProvider(static (x, _) => IsSyntaxTargetForGeneration(x), (x, cancellationToken) => GetSemanticTargetForGeneration(x, cancellationToken))
+                .Where(x => x is not null);
 
-                            // Possible match, now check the string value
-                            string attrName = ca.Name.ToString();
-                            if (attrName == EventSourceAttribute || attrName == EventSourceAttributeShort)
-                            {
-                                // Match add to candidates
-                                _candidateClasses ??= new List<ClassDeclarationSyntax>();
-                                _candidateClasses.Add(classDeclaration);
-                                return;
-                            }
-                        }
-                    }
-                }
-            }
+            context.RegisterSourceOutput(eventSourceClasses, static (spc, x) => EmitSourceFile(spc, x));
```

Comment:
Nit:

```suggestion
            context.RegisterSourceOutput(eventSourceClasses, EmitSourceFile);
```

To avoid the extra method generation.

Reply:
Is the repository already building with the newer version of the compiler that caches this delegate?

Reply:
I forgot that Roslyn started caching the delegate on method groups, I converted the lambdas (even `IsSyntaxTargetForGeneration` where I gave it an unused cancellation token). Even uncached it won't hurt much performance-wise, it's just some delegates in the `Initialize` method.